### ISSUE TITLE
Select multiple photos from camera roll

### ISSUE THREAD

**Issue by bessgreen** [department-of-veterans-affairs/va.gov-team#20994]
## User story
As a user, I want the ability to select multiple photos from my camera roll, so that I don't have to repeat the process of adding photos one-at-at-time. 

## Acceptance Criteria
1. TBD - requirements needed

## Open questions
1. When the camera roll is open, should users be prevented from selecting more than 10 photos? Or, should they be allowed to select >10, but then when they try to upload, they see an error message?
     - If the latter, what should the copy for the error message be?

**Comment by bessgreen** [comment]
@bridgethapner looping you in here! Based on our conversation earlier, it sounds like we'd want to allow the user to multi-select any # of photos from their gallery. If they upload more than they're allotted, then we'll display an error message. 

Follow-up Qs:

1. What should the copy for the error message be? 
2. How should we handle a scenario that incorporates multiple errors? For example: a user tries to upload 20 photos, with unsupported file types, and the total exceeds 50 MB.

**Comment by bessgreen** [comment]
Hi @bridgethapner, @komaly15 finished the corresponding dev spike (#21474) for this ticket. Wanted to loop you in on the findings:

- on iOS, we'll be able to limit the # of photos a user selects directly from the camera roll
     - Example: If a user has uploaded 3 photos to the page, and they open their camera roll, we can limit them to only selecting 7 photos. 
- on Android, we'll need to check the # of photos _after_ the images have been selected from the camera roll
     - This means a user will be able to select >10 images from the camera roll, so we'll need to display an error message 

_________

In addition to questions 1 & 2 in the comment above:

3. for iOS, will we want to limit the # of photos directly from the camera roll? Or should we maintain the same UX across iOS and Android, where a user can select any # of photos from the camera roll and they'll see an error message if they upload more than they're allotted?

**Comment by bridgethapner** [comment]
Hi Bess and Komal! 

Thank you so much for looking into this, it's exciting that it feels possible for app store launch! 

- I think, on iOS, since we can limit the # directly from the camera roll that's what we should do. That seems like the simplest experience. 
- On Android, where in the flow would the error display? For example, would they be back on the preview screen in the app and only the first 10 photos they selected would show + an error message? I am assuming we can't show an error message on the photo selection screen.

**Comment by bessgreen** [comment]
@bridgethapner the error would display on these pages:

![Screen Shot 2021-03-22 at 3.23.16 PM.png](https://images.zenhubusercontent.com/5f64e41291069a617acc1b47/7ca2dee5-dc04-4f12-b9cc-8cd1290f6ad6)

**Comment by bridgethapner** [comment]
@bessgreen Thank you! Will the app successfully have selected & display up to 10 photos and omitted the rest? If that's the case I think we can use the following copy for the error: 

_You can only upload up to 10 files in one submission. Of the files you selected, some were added to your submission. Please review the images below._ 

Re handing multiple errors - I think we should do whatever is simplest for now. Would that be stacking alert components?

**Comment by bessgreen** [comment]
@bridgethapner if a user tries to upload more photos than they're allotted, we'd inevitably want some rule around _which_ photos make it through the gate (is it the first 10 that they selected from the camera roll? what happens if they deselect one, but reselect it later? should the app skip over any files that are too big/wrong format, or should these be included in the upload attempt? etc...). This will add complexity to the flow that might not be worthwhile for MVP, and based on what we've seen w/ the library we're using, it doesn't look like we can guarantee which photos would make it through anyway. 

TL;DR: If a user uploads more photos than they're allotted, the most straightforward approach is to prevent them from uploading any, and ask them to try again by selecting fewer photos. 

___
For the multiple error handling, we can do either of the following (both are around the same LOE):
1. Stack multiple errors (file size, type, too many photos)
2. Display 1 error, where the copy accounts for any combination of errors

Regardless of which one we go with, something to note is that the language of the current error states refers to a single file: "This file cannot be added..." (screenshot below). Now that we're introducing the ability to add multiple files at once, we might want to tweak the language a bit in case they try to add multiple files at once, and 1 or more of the files has issues. This might be as simple as updating it to "One or more files cannot be added because..."
![Screen Shot 2021-03-26 at 9.33.46 AM.png](https://images.zenhubusercontent.com/5f64e41291069a617acc1b47/43ef33ec-7333-4199-a7bb-fb1aa90bd42b)

cc: @komaly15

**Comment by drjecker** [comment]
On the current Secure Messaging website how they handle it is to just display 1 error. So for example when I tried to upload a file that was 1) too large _and_ b) not an accepted file type - it only gave me the `The file type you are trying to upload is not allowed.` error message. I am fine sticking with that here for simplicity sake. There may be other considerations for claims though, since as Bess notes, you have the ability to upload more file at once. SM is a one-at-a-time flow as it current stands.

<img src="https://images.zenhubusercontent.com/602e6bbb345265c801d416d8/7b050218-5050-44cd-9674-d974956f69d3" alt="Screen Shot of error message" width="350">

**Comment by bessgreen** [comment]
@bridgethapner circling back here based on @drjecker's input for SM... one option we hadn't discussed for Claims is to do what Jen's suggesting for SM: if 1 upload triggers multiple errors, just display 1 of the individual errors.

If this is the route we want to take, we'd just need to know the hierarchy for which individual error state should appear first if it's triggered. (In Jen's example above, the 'file type' error trumps the 'file size' error). 

For Claims, the individual errors we're working with - and would need to establish a hierarchy for - are: 
- file type
- file size
- too many photo(s) selected from camera roll _(we haven't built this out yet)_

**Comment by bridgethapner** [comment]
Thanks @bessgreen! 

Re the first point about not allowing any photos to be successfully uploaded if the user selects more than the allotted amount, that makes sense to me - looping in @msmanak here to get her thoughts since she's closest to all things claims file upload right now. 

As far as errors, the MHV example is helpful to see! I think establishing a hierarchy makes sense and is a good direction for MVP but definitely needs some thought & pressure testing since there are lots of factors. The selection # seems like the trickiest part..Meli, is this something you have capacity to help define?
